### ISSUE TITLE
Add `SS_BASE_PROTOCOL` env

### ISSUE THREAD

**Issue by OldStarchy** [silverstripe/silverstripe-framework#10296]
## Affected Version

At least 4.6.2 to current (4.10) but probably all of 4.x. 

## Description

SilverStripe ignores the protocol specified in `SS_BASE_URL`.

## Steps to Reproduce

Help us with step-by-step instructions.
1. Have a server running behind an SSL terminal proxy
2. Set the env var to be the publically facing URL `SS_BASE_URL="https://example.com/"`
3. [`Director::is_https`](
https://github.com/silverstripe/silverstripe-framework/blob/6ced576b2f2d5cf0adcdd8b48f576bde4bd1d745/src/Control/Director.php#L639) uses detects the protocol from the current request instead of using the environment vars.
4. Any urls generated by $resourceURL have the wrong protocol

SilverStripe base URL detection works well, and setting an `SS_BASE_URL` overrides it, but it ignores the protocol. This is annoying because the protocol is the only part that I want to change. The simplest configuration for this would be to add something like `SS_BASE_PROTOCOL` that would let SilverStripe do its thing when working out the domain but let us specify the protocol. This is especially useful when a project is deployed to multiple locations (prod, UAT, review app 1, review app 2, etc., localhost).

On localhost there's no proxy, so plain http is good
On review apps, at least in my setup, i access it via https but the server is still running in http behind the proxy
On prod and UAT the server itself is running via https in most cases, but we have some hosts that use a https proxy.

**Comment by GuySartorelli** [comment]
The protocol in `SS_BASE_URL` isn't ignored by everything in the system - for example it's used in [`Director::baseURL()`](https://github.com/silverstripe/silverstripe-framework/blob/6ced576b2f2d5cf0adcdd8b48f576bde4bd1d745/src/Control/Director.php#L671).

Instead of adding a whole new environment variable, which we'd then have to consider how it interacts with any projects already adding a protocol to `SS_BASE_URL` - a more consistent approach for this specific issue might be to look at where in the sequence it might be appropriate to use the protocol from `SS_BASE_URL` in `Director::is_https()` - though we'd also have to look at where this is used and whether that's likely to break anything.

**Comment by OldStarchy** [comment]
There are two reasons why I suggested a new variable over changing `SS_BASE_URL`;

1. Changing how `SS_BASE_URL` works could be a breaking change.
2. I would still like to rely on SilverStripe to figure out the base URL itself. This is especially important for running sites in dev containers as there's [no way to know what port the site will be exposed on](https://github.com/microsoft/vscode-remote-release/issues/5571#issuecomment-953810837) until it starts serving requests, at which time its too late to update the .env.

**Comment by emteknetnz** [comment]
@OldStarchy have you tried setting the `alternate_base_url` config to https://mysite.com? - https://github.com/silverstripe/silverstripe-framework/blob/6ced576b2f2d5cf0adcdd8b48f576bde4bd1d745/src/Control/Director.php#L630 looks like that has higher precedence than the request in `is_https()`

**Comment by OldStarchy** [comment]
Yes I've been implementing this based on [this StackOverflow answer](https://stackoverflow.com/a/66019407/3094008)

```yml
Silverstripe\Control\Director:
  alternate_base_url: '`SS_BASE_URL`'
```

And this works well in our CI review apps which all sit behind traefik and one project that operates behind a proxy on production, but it's a pain elsewhere.

On our development server, we don't set `SS_BASE_URL` as we have a lot of website instances (150+) and it takes extra effort to set this up when SilverStripe can easily work out the answer for us automatically. And it works because we don't need to override the protocol.

I will change the yml config above to use `_config.php` so that the `alternate_base_url` is not set if `SS_BASE_URL` is also not set. On our dev server the typical base URL looks like `https://dev.exmaple.com/[client]/[project]/develop/[version]/`, if `alternate_base_url` is not set correctly then we start getting redirects to `https://dev.exmaple.com/` and thus the site becomes unusable.

Altering the behavior of `Director::is_https()` to allow overriding from .env without having to specify the full domain and path would efficiently solve this.

**Comment by kinglozzer** [comment]
>I will change the yml config above to use _config.php so that the alternate_base_url is not set if SS_BASE_URL is also not set.

You could also conditionally set that config value with:

```yml
---
Only:
  envvarset: 'SS_BASE_URL'
---
Silverstripe\Control\Director:
  alternate_base_url: '`SS_BASE_URL`'
```

**Comment by sminnee** [comment]
There are a couple of things here.

### Original poster should use `X-Forwarded-...` headers

Reverse proxies can set the `X-Forwarded-Proto: https` header, and if the Silverstripe site lists the proxy's IP in `SS_TRUSTED_PROXY_IPS`, the correct protocol will be inferred. If you do this, and you're only changing the protocol, you probably don't even need SS_BASE_URL. If you're changing the host, there is X-Forwarded-Host as well. 

### Ambiguity in difference between `SS_BASE_URL` and `alternate_base_url`.

In `Director::is_https()`, the current request takes precedence over SS_BASE_URL. The current request does not make use of SS_BASE_URL, and probably shouldn't? The difference between the current URL and SS_BASE_URL can be used for things like domain canonicalisation or SSL enforcement directions. In any case it would be a pretty big change.

The main difference between `Director.default_base_url` (which is set to SS_BASE_URL) and `Director.alternate_base_url` seems to be that alternate_base_url overrides Director::protocolAndHost() and default_base_url does not.

From memory, alternate_base_url was introduced for things like your CMS being hosted on a different domain from your website, and needing to render static-published pages using the front-end domain. (I think we did this for the 2008 DNCC website).

default_base_url is interpreted as "the URL to use when you're outside of a web context" such as CLI/cronjobs. I think this was added much later, because 2008 was before we had yaml-based config.

Notably, that is an obsolete use-case:
 * You could just as easily alter default_base_url in a script that renders static content (this wasn't the case in the version of SS where alternate_base_url first appeared)
 * If you're not using the base tag, it might not be needed at all

So perhaps, we want this:

 * Metadata about the current request (the HTTPRequest in Injector) is based on what the server provides. The main way of overriding this is with X-Forwarded headers from trusted proxies.
 * Absolute URLs are generated from Director.default_base_url, and therefore SS_BASE_URL, if available.
 * alternate_base_url is deprecated as it performs the same role as default_base_url/SS_BASE_URL

This would be achieved by changing Director::host and Director::is_https.
